src/main/java/bio/terra/stairway/FlightMap.java

```diff
@@ -77,33 +102,78 @@ public void makeImmutable() {
    * @throws ClassCastException if found, not castable to the requested type
    */
   public <T> T get(String key, Class<T> type) {
-    Object value = map.get(key);
+    String value = map.get(key);
     if (value == null) {
       return null;
     }
 
-    if (type.isInstance(value)) {
-      return type.cast(value);
+    try {
+      return getObjectMapper().readValue(value, type);
+    } catch (final JsonProcessingException ex) {
+      throw new ClassCastException(
+          "Found value '" + value + "' is not an instance of type " + type.getName());
     }
-    throw new ClassCastException(
-        "Found value '" + value.toString() + "' is not an instance of type " + type.getName());
   }
 
+  /**
+   * Gets the raw (serialized) data stored for an Object in the hash map.
+   *
+   * @param key Key to lookup the data for in the map.
+   * @return String mapped to key, or null if key does not exist.
+   */
+  public String getRaw(String key) {
+    return map.get(key);
+  }
+
+  /**
+   * Serialize an Object and place it into the map at a given key.
+   *
+   * @param key Key to store the object under.
+   * @param value Object to serialize and store.
+   * @throws JsonConversionException if object could not be serialized.
+   */
   public void put(String key, Object value) {
+    try {
+      map.put(key, getObjectMapper().writeValueAsString(value));
+    } catch (final JsonProcessingException ex) {
+      throw new JsonConversionException("Failed to convert object to json string", ex);
+    }
+  }
+
+  /**
+   * Stores a raw (serialized) data value for a given key.
+   *
+   * @param key Key to store the data under.
+   * @param value Data to store.
+   */
+  public void putRaw(String key, String value) {
     map.put(key, value);
   }
 
   public String toJson() {
```

Comment:
Should this be `@Deprecated` now? And `fromJson` also? We won't be converting the whole map to and from json once the transition is complete, right?

Reply:
toJson() will go away sooner, fromJson() will need to stick around until all of the Flights written before this merges are gone.